PFW-1386 Implement M704, M705, M706, and more

gudnimg opened the pull request:
JIRA ticket: https://dev.prusa3d.com/browse/PFW-1386

This PR implements the gcodes: M704, M705, M706, M707, M708, and M709. The PR then also changes the implementation for M701 and M702.


* Internal Parameters don't need to be implemented for the MK3/S printers.
* M702 and M709: I didn't add the `T` parameter since the MK3S only supports one extruder at a time.
* M707 and M708: This PR just created empty functions `ReadRegister()` and `WriteRegister()`. A separate PR will implement these methods.

----

Since the MK3S firmware does not follow the Confluence documentation 100% I want to rewrite the documentation here mainly for myself as a sanity check for what to test.

M701 (load to nozzle)
===

M701 is being changed completely => it is not backwards compatible. Instead of Loading the filament, the firmware will load to the nozzle. This is consistent with the M701 behaviour when the MMU is disabled. The `E` parameter is removed.

M701 is now compatible with Marlin except for the `T` parameter is not used since the team has decided to not support more than one extruder for the MK3S firmware.

New parameters (some are defined by Marlin https://marlinfw.org/docs/gcode/M701.html):
* **P** - _Index of MMU slot (range from 0 to 4)_
* **L** - _Extrude distance for insertion (positive value) (manual reload)_
* **Z** - _Move the Z axis by this distance_

Test cases:
- [x] Issue a M701 P0 --> MMU loads filament in slot 0 to the nozzle and Z-axis moves by `MIN_Z_FOR_LOAD` 
    - 🔴 Printer does not preheat automatically --> cold extrusion! It would be nice if we could abort the MMU command. This has also been an issue without the MMU so this isn't exactly a new problem.
    - ✅ If I preheat beforehand, test passes 
- [x] Issue a M701 P0 Z100 --> MMU loads filament in slot 0 to the nozzle and Z-axis moves by 100mm
    - ✅If I preheat beforehand, test passes
- [x] Issue a M701 P0 Z10 --> MMU loads filament in slot 0 to the nozzle and Z-axis moves by 10mm
    - ✅If I preheat beforehand, test passes
- [x] Repeat the tests above with the MMU disabled
----

M702 (unload filament)
===

M702 parameters are changing.

I threw the `C` parameter out the window. M702 is now compatible with Marlin except for the `T` parameter for the same reasons as in M701.

New parameters (as defined by Marlin https://marlinfw.org/docs/gcode/M702.html):
* **U** - Retract distance for removal (manual reload)
* **Z** - _Move the Z axis by this distance_

- ✅ Issue a M702 --> Unloads filament
- ✅ Issue a M702 Z100 --> MMU unloads filament and Z-axis moves by 100mm
- ✅ Issue a M702 Z10 --> MMU unloads filament and Z-axis moves by 10mm
- [x] Repeat the tests above with the MMU disabled

----

M704 (Load to MMU)
===

For the MMU this does the same thing as M701 previously. Same functionality as "Load filament" from the LCD.

Parameters:
* **P** - _Index of MMU slot (range from 0 to 4)_

- ✅ Issue command M704 --> Nothing happens
- ✅ Issue command M704 P1 --> MMU loads filament from slot 1
- ✅ Issue command M704 P6 --> Nothing happens

----

M705 (Eject filament)
===

Same functionality as "Eject filament" from the LCD.

Parameters:
* **P** - _Index of MMU slot (range from 0 to 4)_

⚠️ Currently not testable. I verified M704 works.

----

M706 (Cut filament)
===

Same functionality as "Cut filament" from the LCD.

Parameters:
* **P** - _Index of MMU slot (range from 0 to 4)_

⚠️ Currently not testable. I verified M704 works.

----

M707 (Read variable)
===

Parameters:
* **A** - _Address of register in hexadecimal_
* **C** - _number of bytes to read_

The address parameter supports both hexadecimal and integers:
* Hexadecimal values must be prepended with `0x` (i.e. `0xFF`)
* Integer values are interpreted  if `0x` is not seen: `255` equal `0xFF`. 
* A hexadecimal value without the `0x` such as `FF` will be interpreted as `0`.

⚠️ Currently not testable

----

M708 (Write variable)
===

At the moment M708 only supports writing one byte. Discussion is needed on how and if we want to handle multiple bytes.

Parameters:
* **A** - _Address of register in hexadecimal_
* **X** - _data to write_
* **C** - _number of bytes to write_

The address parameter supports both hexadecimal and integers:
* Hexadecimal values must be prepended with `0x` (i.e. `0xFF`)
* Integer values are interpreted  if `0x` is not seen: `255` equal `0xFF`. 
* A hexadecimal value without the `0x` such as `FF` will be interpreted as `0`.

⚠️ Currently not testable

----

M709 (MMU reset)
===

Reset the MMU via software or hardware.

Parameters:
* **X** - _0 for software reset, 1 for hardware reset_

Test cases:
- ✅ Issue a M709 X0 --> MMU resets
- ✅ Issue a M709 X1 --> MMU resets
- ✅ Issue a M709 X --> MMU resets
- ✅ Issue a M709 X2 --> Nothing happens

gudnimg committed:
PFW-1386 Implement M704, M705, and M706

gudnimg committed:
PFW-1386 Add M709

DRracer commented on Firmware/Marlin_main.cpp:
```diff
@@ -8763,6 +8763,106 @@ SERIAL_PROTOCOLPGM("\n\n");
+    */
+    case 704:
+    {
+        uint8_t mmuSlotIndex = 0xffU;
```
it looks like this part is common to all M704, 5, 6, it could be extracted to a common function later to save space

gudnimg replied:
I don't see any easy way to create a function pointer. C++ syntax seems overly complicated :( 

I have a common function setup like this:

```c
static void gcodes_M704_M705_M706(bool (*func)(uint8_t))
{
    uint8_t mmuSlotIndex = 0xffU;
    if (MMU2::mmu2.Enabled() && code_seen('P'))
    {
        mmuSlotIndex = code_value_uint8();
        if (mmuSlotIndex < MMU_FILAMENT_COUNT) {
            func(mmuSlotIndex);
        }
    }
}
```

The problems seems to be creating a the pointer to the function method. Its not as trivial as normal C it seems.

gudnimg replied:
Here is a less clean solution but it saves 36 bytes of flash
```c
static void gcodes_M704_M705_M706(uint16_t gcode)
{
    uint8_t mmuSlotIndex = 0xffU;
    if (MMU2::mmu2.Enabled() && code_seen('P'))
    {
        mmuSlotIndex = code_value_uint8();
        if (mmuSlotIndex < MMU_FILAMENT_COUNT) {
            switch (gcode)
            {
            case 704:
                MMU2::mmu2.load_filament(mmuSlotIndex);
                break;
            case 705:
                MMU2::mmu2.eject_filament(mmuSlotIndex, false);
                break;
            case 706:
                MMU2::mmu2.cut_filament(mmuSlotIndex);
                break;
            default:
                break;
            }
        }
    }
}
```

gudnimg committed:
PFW-1386 Add M707 and M708

WriteRegister and ReadRegister function will create new compiler
warnings due to unused parameters, we can ignore it for now.

gudnimg committed:
PFW-1386 Implement changes to M701

gudnimg committed:
PFW-1386 Raising Z-axis during M701 is now done via Gcode

If the Z-parameter is not issued, then default behavior will raise
the Z-axis by MIN_Z_FOR_LOAD

gudnimg committed:
PFW-1386 Implement changes to M702

gudnimg committed:
PFW-1386 if automatic = false, raising Z-axis is done via gcode

gudnimg commented:
Gcodes are all implement, now I just need to test my changes and optimise code size.

gudnimg committed:
PFW-1386 Create a common function for M704 to M706

Saves 36 bytes of flash

gudnimg commented:
Current usage after [e866326](https://github.com/DRracer/Prusa-Firmware/pull/66/commits/e8663266974d52dcd5bc600b15e36f3cc2873e4a)
```
Sketch uses 250036 bytes (98%) of program storage space. Maximum is 253952 bytes.
Global variables use 6021 bytes of dynamic memory.
```

gudnimg committed:
PFW-1386 M701 now loads to nozzle with MMU

gudnimg commented:
Hit my first issue in testing. M701 will load to the nozzle when the the nozzle is cold --> cold extrusion.

This is what Marlin does:
1. Enter `load_filament_to_nozzle()`
2. Check if the temperature is not too cold (175°C or `extrude_min_temp`)
3. If its too cold, the printer beeps and raises an error message on the LCD.
4. Else the load continues

If the temperature is too cold we need to be able to stop the MMU from pushing more and more filament into the extruder. We need to be able to abort the "load to nozzle" command. I'm not sure if that is currently possible.

gudnimg committed:
PFW-1386 Implement 'L' parameter for M701

It is only used when the MMU is disabled. With the MMU we use a defined ramming sequence.

When the MMU is disabled I borrowed the usages from Marlin 2.
The L parameter only defines the length for the "fast" load length.
Default value is 70mm, same as M600

gudnimg committed:
PFW-1386 Implement 'U' parameter for M702

Similar to 'L' in M701, this does not apply to the MMU
The U parameter controls how much mm the extruder will unload
Default value is 80mm

gudnimg commented:
'L' and 'U' parameters are implemented now. Surprisingly the Flash usage has dropped a little bit (by 46 bytes):

```
Sketch uses 249990 bytes (98%) of program storage space. Maximum is 253952 bytes.
Global variables use 6021 bytes of dynamic memory.
```

gudnimg committed:
PFW-1386 Optimise default value assign z_target

gudnimg committed:
PFW-1386 I don't see a reason for these lines

So I'm removing them

gudnimg committed:
PFW-1386 Address should accept hexidecimal values

Now these are equal:
M707 A255 C1
M707 A0xFF C1

Both address inputs are interpreted as FF

gudnimg commented:
Rebased the branch with mk3-mmu.

mk3-mmu memory:
```
Sketch uses 248918 bytes (98%) of program storage space. Maximum is 253952 bytes.
Global variables use 5670 bytes of dynamic memory.
```

This PR:
```
Sketch uses 249228 bytes (98%) of program storage space. Maximum is 253952 bytes.
Global variables use 5670 bytes of dynamic memory.
```
So far the PR only adds 310 bytes of flash

gudnimg committed:
PFW-1386 Minor correction to M707 and M708 useage example

gudnimg committed:
PFW-1386 Remove old debug message to save space

Saves 34 bytes of flash

gudnimg commented:
There is one issue with the Z prameter. I’ve been testing on an unhomed printer. `raise_z_above` will not work properly for every position. If the Z input is lower than the current position (such as during a print) the Z parameter input will be ignored. 

I will try to correct this, but it will likely increase flash usage...

For future reference this is what the memory consumption looks like now:

```
Sketch uses 249194 bytes (98%) of program storage space. Maximum is 253952 bytes.
Global variables use 5670 bytes of dynamic memory.
```

TODO:
Memory consumption after:

gudnimg committed:
PFW-1386 M701, M702: Move Z-axis using relative position

* Created a new function raise_z() which will move the Z axis by some
X millimeters. If plan = false, the function will return the
actual travel distance since the move is blocking.
* raise_z_above() is refactored to call raise_z()

Now the M701 and M702 will move the Z-axis relatively, and also
when done, revert the Z axis position when done.
This is a similar behavior as in Marlin 2.

gudnimg committed:
PFW-1386 Optimisation: st_synchronize is not needed after raise_z_above

Saves 16 bytes of flash

gudnimg committed:
PFW-1386 fixup

gudnimg commented:
It seems I've broken the unload sequence with the MMU disabled. We need some sort of ramming sequence to prevent stringing.

gudnimg committed:
PFW-1386 Use same code for Z-lift in M600 as in M701/M702

This reduces flash usage by 128 bytes!

gudnimg committed:
PFW-1386 Use Marlin 2 unload sequence for single material

The current unload sequence does not give good results.
The Marlin 2 sequence looks good on my end. This can be later
used to improve M600

Added all the constants from Marlin 2

gudnimg commented:
@DRracer I realise I'm probably going out of PFW1386's scope with this change [ba12838](https://github.com/DRracer/Prusa-Firmware/pull/66/commits/ba128384c432e82270bac2d4dbb799a49076748f) let me know if you want the commit dropped.

If this commit is dropped, then we need to do something about the unload sequence. I think there is a lot of room for improvement in M600, but I don't like to touch it in this PR as this PR is already quite big.

Current memory consumption:
```
Sketch uses 249326 bytes (98%) of program storage space. Maximum is 253952 bytes.
Global variables use 5670 bytes of dynamic memory.
```

PR adds 408 bytes of flash, 0 bytes of RAM

DRracer commented:
@gudnimg I think the extent of this PR is ok. For now please keep the ba12838 commit there, we are getting reports about potentially bad unload sequence/ramming anyway.

I'll go through your PR tomorrow with a fresh brain, stay tuned :wink:

DRracer commented on Firmware/Marlin_main.cpp:
```diff
@@ -3817,14 +3800,13 @@ static void gcode_M600(bool automatic, float x_position, float y_position, float
         }
-    
+
+        // Recover Z axis
```
we shall first return XY and then move Z down - it's been that way before for a reason

gudnimg replied:
Moving the Z after XY wasn’t working in my testing with `raise_z` which is why I moved the Z before XY, then it suddenly worked.

I think it has something to do with the XY move using `current_position[Z_AXIS]` because using raise_z after the XY move, no movement happens.

DRracer replied:
:thinking: this doesn't sound right, there might be some other problem hidden in the back.
At least let's add a comment about this issue, because we'll run into it again later.

gudnimg replied:
Ok I’ll add a TODO in a few minutes :)

gudnimg replied:
Done

DRracer commented on Firmware/Marlin_main.cpp:
```diff
@@ -3843,31 +3825,23 @@ static void gcode_M600(bool automatic, float x_position, float y_position, float
+        current_position[E_AXIS] += fastLoadLength;
         plan_buffer_line_curposXYZE(400 / 60); // fast sequence
         st_synchronize();
 
```
Why was the Z-move removed? Or is it done somewhere else?

gudnimg replied:
Yes it is now done in the M701 gcode with `raise_z`. Now it applies to both single material and MMU.

DRracer commented on Firmware/Marlin_main.cpp:
```diff
@@ -8721,45 +8715,217 @@ SERIAL_PROTOCOLPGM("\n\n");
+
+    #### Parameters
+    - `A` - Address of register in hexidecimal. Default value is 0.
+    - `C` - Number of bytes to read. Default value is 0.
```
We will probably drop the `C` parameter - the MMU protocol only transfers 16bit integers, so reading/writing more or less bytes is not supported.
I will do that in the following PR which adds the read/write capabilities.

DRracer commented on Firmware/Marlin_main.cpp:
```diff
@@ -8721,45 +8715,217 @@ SERIAL_PROTOCOLPGM("\n\n");
+                addr = uint8_t(strtol(strchr_pointer+1, NULL, 0));
+            }
+            if( code_seen('X') ) {
+                data = code_value_uint8();
```
this will become code_value_uin16()

DRracer commented on Firmware/Marlin_main.cpp:
```diff
@@ -8721,45 +8715,217 @@ SERIAL_PROTOCOLPGM("\n\n");
+        M708 A0x14 X30 C1 - Write to register 0x14 the value 30 which is 1 byte.
+
+    #### Parameters
+    - `A` - Address of register in hexidecimal. Default value is 0.
```
address register should be mandatory, address 0 is major firmware version i.e. a read only register :wink:

DRracer reviewed (approved):
Very nice :+1: 
There a few remarks related to the read/write register functionality, but that will be done in a separate PR.

gudnimg commented:
For what its worth here is a 6 hour single material print (with MMU enabled) I had going over the night. Went very smoothly. (This of course doesn’t cover all changes in this PR) :)

![BFDD5684-10FA-42CD-8138-87F3E337FD8E](https://user-images.githubusercontent.com/8218499/187139696-6bea7899-f845-43d2-b560-015ac3c59ac5.jpeg)
![2CE16D34-DB5F-4F12-8675-68CE9528D944](https://user-images.githubusercontent.com/8218499/187139715-eff79e16-151c-4cfb-a646-f0ace04d448e.jpeg)

gudnimg committed:
PFW-1386 Add a TODO

When recovering the Z-axis in M600, we'd like the Z move to
happen after the XY move.

gudnimg commented:
rebased the branch with respect to `mk3-mmu` to take in the filament sensor optimisations

Current stats:
```
Sketch uses 249082 bytes (98%) of program storage space. Maximum is 253952 bytes.
Global variables use 5654 bytes of dynamic memory.
```

gudnimg committed:
PFW-1386 Clamp Z-axis when homed

gudnimg committed:
PFW-1386 Refactor raise_z

Removed the "plan" parameter.
We were incorrectly returning from the function if the printer was homed,
but plan = false. This would leave current_position with an incorrect value

If the printer is homed => finish the move and return the travel distance

If the printer is not homed => rely on end stop to prevent damage,
return travel distance even if the endstop stopped the move.

gudnimg commented:
@DRracer @3d-gussner Small update to the PR

* [2061afb](https://github.com/DRracer/Prusa-Firmware/pull/66/commits/2061afb2460341bf025808c27ddc06b70f3342ec) Fixes an issue where if the printer was homed, the Z-axis could crash because the axis was not clamped.
* [dcce520](https://github.com/DRracer/Prusa-Firmware/pull/66/commits/dcce520b7633631a7b89adb8d26f0395a7ed3d78) Fixes an issue with `plan` parameter. Its removed now. If the printer is homed we will now just finish raising the Z-axis before leaving the function. Just like when the printer is not homed. I believe this is safer.

gudnimg committed:
PFW-1386 st_get_position_mm() is more accurate when using endstops (not homed printer)

If the endstops kick in, then st_get_position_mm shows the actual travel distance. current_position[Z-AXIS] does not get updated correctly thus we cannot use it when unhomed.

current_position[Z-AXIS] is later set to Z_MAX_POS,
we cannot use that either to calculate the travel distance

gudnimg commented:
Current stats:

```
Sketch uses 249096 bytes (98%) of program storage space. Maximum is 253952 bytes.
Global variables use 5654 bytes of dynamic memory.
```

3d-gussner reviewed (approved):
Approved but there are few things we need to fix in next PR

- M704 and LCD Load filament goes in a endless loop as soon filament is loaded to the nozzle
- M707 and LCD Cut filament should not be executed/shown when LCD Settings --> Cutter [OFF]
